src/backend/farmer/maybe_node_client.rs

```diff
@@ -1,127 +1,708 @@
+                            }
+                        };
+
+                    cached_archived_segment
+                        .lock()
+                        .replace(CachedArchivedSegment::Weak(Arc::downgrade(
+                            &archived_segment,
+                        )));
+
+                    maybe_archived_segment_header
+                };
+
+                Box::pin(async move { maybe_archived_segment_header })
+            });
+
+        let archived_segment_acknowledgement_senders =
+            self.archived_segment_acknowledgement_senders.clone();
+        let fut = async move {
+            let mut archived_segment_acknowledgement_senders =
+                archived_segment_acknowledgement_senders.lock();
+
+            archived_segment_acknowledgement_senders
+                .senders
+                .remove(&subscription_id);
+        };
+
+        self.subscription_executor.spawn(
+            "subspace-archived-segment-header-subscription",
+            Some("rpc"),
+            fut.boxed(),
+        );
+
+        Ok(Box::pin(stream))
     }
 
-    async fn segment_headers(
+    async fn acknowledge_archived_segment_header(
         &self,
-        segment_indexes: Vec<subspace_core_primitives::SegmentIndex>,
-    ) -> Result<Vec<Option<subspace_core_primitives::SegmentHeader>>, Error> {
-        match &*self.inner.load() {
-            Some(inner) => inner.segment_headers(segment_indexes).await,
-            None => Err("Inner node client not injected yet".into()),
+        segment_index: subspace_core_primitives::SegmentIndex,
+    ) -> Result<(), Error> {
+        let archived_segment_acknowledgement_senders =
+            self.archived_segment_acknowledgement_senders.clone();
+
+        let maybe_sender = {
+            let mut archived_segment_acknowledgement_senders_guard =
+                archived_segment_acknowledgement_senders.lock();
+
+            (archived_segment_acknowledgement_senders_guard.segment_index == segment_index)
+                .then(|| {
+                    let last_key = *archived_segment_acknowledgement_senders_guard
+                        .senders
+                        .keys()
+                        .next()?;
+
+                    archived_segment_acknowledgement_senders_guard
+                        .senders
+                        .remove(&last_key)
+                })
+                .flatten()
+        };
+
+        if let Some(sender) = maybe_sender {
+            if let Err(error) = sender.unbounded_send(()) {
+                if !error.is_closed() {
+                    warn!("Failed to acknowledge archived segment: {error}");
+                }
+            }
         }
+
+        debug!(%segment_index, "Acknowledged archived segment.");
+
+        Ok(())
     }
 
     async fn piece(
         &self,
-        piece_index: subspace_core_primitives::PieceIndex,
+        requested_piece_index: PieceIndex,
```

Comment:
It is better to keep argument names the same as in original trait, sometimes helps with code search and generally with consistency.